Implementation of Flask-Sock based WebSocket route

alecpm opened the pull request:
## Description

This is a PR to replace the use of the obsolete `Flask Sockets` library with `Flask-Sock`. Initially it's just a simple port of the current implementation in `experiment_server/sockets.py`. However, the current implementation is completely un-documented, and could probably use some additional features.

A large part of this work will be to document the current WebSocket implementation and how it can be used by experiment developers to provide realtime communication. Additionally, it will probably make sense to extend the integration in various ways make it easier for experiment developers to use these features.

### Changes so far:

* The new WebSocket Client object no longer provides a `closed` attribute, but instead provides a `connected` attribute. Internal checks and tests have been updated to use `connected`.
* The `simple_websocket.Server` class provides a native Ping/Pong mechanism to detect disconnected clients. This allowed for the removal the custom `heartbeat` implementation.
* Error handling has been updated to handle `ConnectionClosed` events from `simple_websocket` in addition to the already handled `socket.error`. As was the case for `socket.error`, a `ConnectionClosed` event will result in the client being unsubscribed from the channel.
* Dependencies have been updated using `scripts/update_dependencies.sh` to pull in `Flask-Sock`. Is there a better way to do this? I'd like to make sure we're not breaking backwards compatibility with these dependency updates which go well beyond the scope of the WebSocket support in this PR.
* Added a `publish_to_subscribers()` method to the experiment class that simplifies publishing messages to a specific pub-sub channel.
* Added a control message channel (`dallinger_control`) where messages regarding client connections/disconnections and channel subscribe/unsubscribe events are sent.
* Added a `client_info` method to the `socket.Client` and `Experiment` classes to provide info about the specific client in control messages.
* Allow for the `worker_id` and `participant_id` to be sent as arguments the `/chat` route, which will then be included in the `client_info` data for `socket.Client`.

### To Do:

* [ ] Update documentation
* [x] Implement event/messaging around client connect and disconnect to allow experiments to provide custom handling when such events occur.</strike>
* [ ] Additional Experiment integration
  * [x] Add experiment method to publish messages to clients.
  * [ ] ...

### Summary of current functionality:

* The experiment server includes a `/chat` WebSocket route that takes a `channel` request argument to specify a "Chat" `Channel` name. Each channel is backed by a Redis pubsub channel.
* Clients that make a WebSocket Connection to the `/chat` route are subscribed as Clients to that "Chat" channel. Any message sent by the client is published to the Redis pubsub channel. All messages published to the channel are sent to all subscribing Clients.
* Messages are formatted as `channel_name:payload` where the payload is generally a JSON string (though that's not required). Including the channel name in the message may seem redundant, but it allows for messages to be sent to multiple channels over a single WebSocket connection created by a request to a channel specific `/chat` route.
* An experiment can define a `channel` attribute to indicate a "Chat" `Channel` name that it wants to be a Client for.
  * When such an experiment's `/launch` route is called, the experiment will be subscribed as a Client to the named channel.
  * Whenever messages are published on the channel, the Experiment's `send` method is called with the message as the only argument.

This allows for a good deal of flexibility, but perhaps not enough. For example an experiment might want to subscribe to more than a single channel.  I still wonder if we might be better off using [Flask `Signals`](https://flask.palletsprojects.com/en/2.3.x/signals/) instead of reusing the Redis pubsub channel for these control messages. However, keeping the all socket message handling in a single place seems like a good idea to me.

## Motivation and Context

See issue #5330.

## How Has This Been Tested?

Existing tests of the `/chat` route and backend have been updated to work with the new implementation. New tests have been created for the control messaging, and the experiment and experiment_server integration.

## Screenshots (if appropriate):

pmcharrison commented on constraints.txt:
```diff
@@ -4,14 +4,18 @@
 #    ./scripts/update_dependencies.sh
 #
-
+-e file:///Users/alecmitchell/Development/bundles/ucb_wallace/Dallinger
```
I think you need to have Dallinger installed not in editable mode when generating constraints.

alecpm replied:
It looks like this was an issue with the `update_dependencides.sh` script where the `sed` command which was intended to remove that line (along with all other sed commands) was not working as intended on MacOS. I've fixed the script and re-run it.

pmcharrison commented:
> Dependencies have been updated using scripts/update_dependencies.sh to pull in Flask-Sock. Is there a better way to do this? I'd like to make sure we're not breaking backwards compatibility with these dependency updates which go well beyond the scope of the WebSocket support in this PR.

I think that's right, yes. Normally we execute this script in advance of making a new release.

alecpm commented:
@silviot I'd like it if you could review this to see if the approach makes sense (in particular the use of the `dallinger_control` channel for sending signals related connections/subscriptions). For the most part I'm just making some small tweaks to the existing implementation along with updating the libraries, but I've also added those control messages and a simple method on the experiment class for sending messages to subscribers.

pmcharrison commented on dallinger/experiment_server/sockets.py:
```diff
@@ -114,14 +147,23 @@ def unsubscribe(self, client):
+            "worker_id": self.worker_id,
+            "participant_id": self.participant_id,
+        }
+
```
Do you think we should check that the `worker_id` and `participant_id` are a valid combination (by looking at the database)? Might be good for security?

alecpm replied:
We don't do that sort of thing in other parts of Dallinger, so I'm not sure we need to do so here here. My thought was that this was mostly and either/or situation where `participant_id` would generally be preferred, but if for some reason the connection was established before a `participant_id` was assigned then the experiment might be able to make use of the `worker_id` instead. This information isn't used in Dallinger by default, it would be up to the experiment developer to decide how they might want to use it.

jessesnyder reviewed (approved):
This solution looks great to me. Experience has shown that it's very hard to predict future use-cases in Dallinger, and so I like the flexibility of a relatively generic pub/sub system with channels. I think a demo that shows basic communication with channels (possibly named as "rooms", since that sounds like the most obvious use-case) will effectively communicate how the system is used.

pmcharrison commented on dallinger/experiment.py:
```diff
@@ -291,15 +298,52 @@ def is_overrecruited(self, waiting_count):
-            'shopping:{"type":"buy","color":"blue","quantity":"2"}'
+            ``'shopping:{"type":"buy","color":"blue","quantity":"2"}'``
+
+        control messages about channel subscription and websocket
```
Should 'control' be capitalized?

pmcharrison commented on dallinger/experiment.py:
```diff
@@ -291,15 +298,52 @@ def is_overrecruited(self, waiting_count):
+        be included in control channel messages.
+        """
+        return {
+            "class": self.__class__.__module__ + "." + self.__class__.__module__,
```
Should this be 

```py
 self.__class__.__module__ + "." + self.__class__.__name__,
```

alecpm replied:
Indeed it should!

alecpm commented:
@pmcharrison @jessesnyder @silviot I've added a `chatroom_ws` demo which is simple a clone of the `chatroom` demo that uses WebSockets for communication and delegates the Info/Transmission creation to the experiment class. The clients use only a single chat channel (i.e. `chatroom`) for communication. However, the experiment listens on both that channel and the control channel, and it relays information from the control channel messages back to the `chatroom` channel for the clients to see and potentially act on (for example disabling the chat once all other participants have left the room). I've also added some very basic documentation.

There's also a small fix to make the waiting room work with the `entryInformation` changes we made for #2464.

alecpm commented:
This has been updated to use the concepts from #5407 to run message handling in asynchronous workers. The chatroom_ws demo has been updated to use the new API and process messages asychronously.

jessesnyder commented on dallinger/experiment_server/worker_events.py:
```diff
@@ -109,66 +149,71 @@ def worker_function(
+        self.details = details
 
     @property
     def data(self):
```
Can we add `receive_time` and `details` to the `data` property? (`details` is just data, right?)

alecpm replied:
@jessesnyder I don't see why not, they'd still be instance attributes, right? What is the `data` property used for?

alecpm replied:
Generally `details` should be JSON-ish data that could get stored into the `details` JSONB field on an `Event` object in the DB (as it is already done in the `TrackingEvent` special case), but it could be used for whatever by one of these custom event handlers.

alecpm replied:
I've added those attributes to the data property.

jessesnyder reviewed (changes requested):
This looks great... adds an important foundation for scaling websocket-based experiments and gives the under-performing WorkerEvent classes a legitimate new purpose.

pmcharrison reviewed (commented):
Thanks @alecpm, looking great!

alecpm commented on docs/source/using_websockets.rst:
```diff
@@ -0,0 +1,33 @@
+:attr:`~dallinger.experiment.Experiment.channel` attribute which will subscribe
+the experiment class to a the named channel. Such experiments should implement a
+:func:`~dallinger.experiment.Experiment.receive_message` method to recieve and
+process any messages on the specifed channel as well as the control channel
```
Typos fixed.

jessesnyder reviewed (approved):
Wow, this is a whopper. Looks great!

jessesnyder commented:
@alecpm I'd be in favor of making this test less specific:
tests/test_experiment_server.py::TestAssignmentSubmitted::test_calls_on_assignment_submitted_to_recruiter (`assert_called_once() probably fine).
